### ISSUE TITLE
multiprocessing.Queue.get(block=False) can raise queue.Empty on non-empty queue

### ISSUE THREAD

**Issue by 0acbaada-cdd2-4efe-82bc-1f2d975bdb9b** [python/cpython#87302]
BPO | [43136](https://bugs.python.org/issue43136)
--- | :---
Nosy | @sbraz, @JulienPalard
Files | <li>[queue_test.py](https://bugs.python.org/file49791/queue_test.py "Uploaded as text/plain at 2021-02-05.12:03:05 by @sbraz")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-02-05.12:03:06.227>
labels = ['type-bug', '3.8', '3.9', '3.10', '3.7', 'library']
title = 'multiprocessing.Queue.get(block=False) can raise queue.Empty on non-empty queue'
updated_at = <Date 2021-02-05.15:20:21.144>
user = 'https://github.com/sbraz'
```

bugs.python.org fields:
```python
activity = <Date 2021-02-05.15:20:21.144>
actor = 'mdk'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2021-02-05.12:03:06.227>
creator = 'sbraz'
dependencies = []
files = ['49791']
hgrepos = []
issue_num = 43136
keywords = []
message_count = 2.0
messages = ['386525', '386528']
nosy_count = 2.0
nosy_names = ['sbraz', 'mdk']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue43136'
versions = ['Python 3.6', 'Python 3.7', 'Python 3.8', 'Python 3.9', 'Python 3.10']
```

</p></details>

**Comment by hyperfraise** [comment]
I have the same issue but with block=True and timeout=1

**Comment by CorvetteCole** [comment]
I can also reproduce this with block=False

**Comment by MorningLightMountain713** [comment]
Yup - same behaviour. qsize returns a items in the queue, but get.no_wait raises queue.Empty.

My workaround for this is to block, with a smaill delay in a loop, then it seems to work

```python
        while True:
            try:
                queue.get(timeout=0.05)
            except Empty:
                pass

                size = queue.qsize()

                log.info("Queue length: %s", size)

                if not size:
                    break
```

**Comment by x42005e1f** [comment]
The problem is too simple (and yes, it is also reproducible on any version of Python regardless of platform, and now I will explain why). If we add [`traceback.print_exc()`](https://docs.python.org/3/library/traceback.html#traceback.print_exc) to the `except` block, we will see that the cause in the implementation is [the following block](https://github.com/python/cpython/blob/db098a475a47b16d25c88d95dbcf0c6572c68576/Lib/multiprocessing/queues.py#L106-L107) ("rlock" is read as "read lock"; the name is somewhat ambiguous):

```python
if not self._rlock.acquire(block, timeout):
    raise Empty
```

The meaning is as follows: if someone is already performing the get operation, and because of this we did not have time to acquire the lock to ensure exclusive access, we raise [`queue.Empty`](https://docs.python.org/3/library/queue.html#queue.Empty). In general, this is usually desirable behavior for non-blocking operations, but the current exception is misleading, and neither the [queue](https://docs.python.org/3/library/queue.html) module nor even the [threading](https://docs.python.org/3/library/threading.html) module do this (all underlying lock acquisitions in all their primitives are unconditionally blocking).

**Comment by x42005e1f** [comment]
One solution would be to add a second semaphore that would synchronize get operations (similar to the one that synchronizes put operations). Its initial value would be zero. The put operation releases the semaphore, and the get operation acquires the semaphore (with `block` and `timeout` passed). If the semaphore cannot be acquired during this time, this is only possible if there are no items in the queue, which means we raise [`queue.Empty`](https://docs.python.org/3/library/queue.html#queue.Empty). Otherwise, we unconditionally acquire the lock and perform the operation in exclusive mode.

I implemented something similar in my old queue implementation, so it should work as expected. There are some peculiarities with the order in which items are retrieved, but hardly anyone will notice.

**Comment by x42005e1f** [comment]
A quick workaround in line with the above:

```python
import multiprocessing
import multiprocessing.queues

from queue import Empty


class _GH87302Queue(multiprocessing.queues.Queue):
    def __init__(self, maxsize=0, *, ctx):
        super().__init__(maxsize, ctx=ctx)
        self.__getsem = ctx.Semaphore(0)

    def __getstate__(self):
        return (*super().__getstate__(), self.__getsem)

    def __setstate__(self, state):
        super().__setstate__(state[:-1])
        self.__getsem = state[-1]

    def put(self, obj, block=True, timeout=None):
        super().put(obj, block, timeout)
        self.__getsem.release()

    def get(self, block=True, timeout=None):
        if self._closed:
            raise ValueError(f"Queue {self!r} is closed")

        if not self.__getsem.acquire(block, timeout):
            raise Empty

        return super().get()

    # optional (see the `full()` implementation)
    def empty(self):
        return self.__getsem._semlock._is_zero()


def GH87302Queue(maxsize=0):
    return _GH87302Queue(maxsize, ctx=multiprocessing.get_context())
```

And if you replace

```python
queue_class = multiprocessing.Queue
```

with

```python
queue_class = GH87302Queue
```

in the reproducer script, everything works as expected.

---

Please note that [`multiprocessing.Queue`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Queue), contrary to the documentation, is not actually a class!

```python
>>> import multiprocessing
>>> multiprocessing.Queue
<bound method BaseContext.Queue of <multiprocessing.context.DefaultContext object at 0x7fe870fe36b0>>
```

In the reproducer script, it would be more correct to call it a factory rather than a class.
